delivery_ups_oca/models/ups_request.py

```diff
@@ -0,0 +1,415 @@
+        NumOfPieces = picking.number_of_packages
+        PackageWeight = picking.shipping_weight
+        if is_package:
+            NumOfPieces = sum(package.mapped("quant_ids.quantity"))
+            PackageWeight = max(package.shipping_weight, package.weight)
+            package = package.package_type_id
+        return {
+            "Description": package.name,
+            "NumOfPieces": str(NumOfPieces),
+            "Packaging": {
+                "Code": package.shipper_package_code,
+                "Description": package.name,
+            },
+            "Dimensions": {
+                "UnitOfMeasurement": {"Code": self.package_dimension_code},
+                "Length": str(package.packaging_length),
+                "Width": str(package.width),
+                "Height": str(package.height),
+            },
+            "PackageWeight": {
+                "UnitOfMeasurement": {"Code": self.package_weight_code},
+                "Weight": str(PackageWeight),
+            },
+        }
+
+    def _partner_to_shipping_data(self, partner, **kwargs):
+        """Return a dict describing a partner for the shipping request"""
+        return dict(
+            **kwargs,
+            Name=(partner.parent_id or partner).name,
+            AttentionName=partner.name,
+            TaxIdentificationNumber=partner.vat,
+            Phone=dict(Number=partner.phone or partner.mobile),
+            EMailAddress=partner.email,
+            Address=dict(
+                AddressLine=[partner.street, partner.street2 or ""],
+                City=partner.city,
+                StateProvinceCode=partner.state_id.code,
+                PostalCode=partner.zip,
+                CountryCode=partner.country_id.code,
+            ),
+        )
+
+    def _label_data(self):
+        res = {"LabelImageFormat": {"Code": self.file_format}}
+        # According to documentation, we need to specify sizes in some formats
+        if self.file_format != "GIF":
+            res["LabelStockSize"] = {"Height": "6", "Width": "4"}
+        return res
+
+    def _prepare_create_shipping(self, picking):
+        """Return a dict that can be passed to the shipping endpoint of the UPS API"""
+        if self.use_packages_from_picking and picking.package_ids:
+            # modelo: stock.quant.package
+            packages = [
+                self._quant_package_data_from_picking(package, picking, True)
+                for package in picking.package_ids
+            ]
+        else:
+            # modelo: stock.package.type
+            packages = []
+            package_info = self._quant_package_data_from_picking(
+                self.default_packaging_id, picking, False
+            )
+            package_weight = round(
+                (picking.shipping_weight / picking.number_of_packages), 2
+            )
+            for i in range(0, picking.number_of_packages):
+                package_item = package_info
```

Comment:
shouldn't this be `package_item = package_info.copy()` to ensures modifications to one package don't affect others?

Reply:
You are right, thank you